Add support for reading from stdin

zacchiro opened the pull request:
Given the Beancount loader does not support reading from file-like objects, to
support reading from stdin we write all of it to a named temporary file and
load from the corresponding path name.

Fixes #13

tbm reviewed (changes requested):
Thanks so much!

Can you open a `## 1.4 (unreleased)` entry in `docs/changelog.md` and add a bullet point for this.  If not, I'll do it.

If you re-submit, please ucfirst() the first word of the Subject ;)

tbm commented:
I think the long line also introduces a pylint3 warning but I should look at black as you suggested recently

zacchiro committed:
Add support for reading from stdin

Given the Beancount loader does not support reading from file-like objects, to
support reading from stdin we write all of it to a named temporary file and
load from the corresponding path name.

Fixes #13

zacchiro commented:
> Can you open a `## 1.4 (unreleased)` entry in `docs/changelog.md` and add a bullet point for this. If not, I'll do it.
> If you re-submit, please ucfirst() the first word of the Subject ;)

both done in this version!

tbm commented:
Thanks!

dnicolodi commented on beancount2ledger/cli.py:
```diff
@@ -68,11 +71,24 @@ def cli():
+            tmpfile = stack.enter_context(
+                NamedTemporaryFile(prefix="tmp.beancount2ledger.",
+                                   mode="w",
+                                   encoding=locale.getpreferredencoding()))
```
Beancount input files must be encoded in UTF8. Thus I suspect this is not strictly correct (at least on systems where the environment encoding is not UTF8).

zacchiro replied:
Oh, thanks, I didn't know that. I'll submit a separate pull-request to fix the encoding to utf-8 (and I'll be sorry about that).

dnicolodi commented on beancount2ledger/cli.py:
```diff
@@ -68,11 +71,24 @@ def cli():
+        # stdin we write its content to a tempfile and call the loader on its path
+        if in_file == "-":
+            tmpfile = stack.enter_context(
+                NamedTemporaryFile(prefix="tmp.beancount2ledger.",
```
I haven't tested it, but I suspect this does not work on Windows. There it is not possible to use the name of the NamedTemporaryFile to re-open the file for writing (see for the example the beancount tests failures on Windows). My understanding is that this is because NamedTemporaryFile keeps the file descriptor open and Windows does not allow to re-open the file in this case. Closing the file descriptor causes the NamedTemporaryFile to be deleted.

Also, I think, in the case of processing through a temporary file, it makes sense to explicitly disable beancount caching of the loaded data structures as it causes useless serialization and writes to disk. IIRC there is an environment variable to disable the cache.
